### PR TITLE
fuse: re-use rcat to support uploads for all remotes

### PR DESCRIPTION
**Pull request by breunigs**
With this patch, I can write successfully to an encrypted remote on B2, which was not possible previously. In this case it spools the files, of course, but direct streaming also works fine (tested with Dropbox).

I'm a bit weary to merge this right away, because I suspect 1.38 will be around the corner soon. Having `rcat` exposed to a wider audience might be good to fix any additional bugs before accidentally breaking the mount. There's also the issue that all the business-oriented remotes still need to be checked/implemented (#1614). Previously it would just `Put`, but now it will spool, which might be worse. It will be a while before I get a toy credit card to make test accounts with these providers.

My suggestion would be to merge this for 1.39.

### CODE REVIEW

**Comment by ncw** [comment]
That is brilliant Stefan!

The milestone says 1.38 at the end of the month.  This is a nice feature though so it would be great to get it in...  In my experience stuff doesn't really get tested until it is merged to master...

> There's also the issue that all the business-oriented remotes still need to be checked/implemented (#1614). Previously it would just Put, but now it will spool, which might be worse. It will be a while before I get a toy credit card to make test accounts with these providers.

I have accounts on all the providers so I can help here. I'll can get you a free [memstore (swift)](https://www.memset.com/cloud/storage/) account - I'll email with details.

File: fs/operations.go [hunk 1]
```diff
@@ -1583,7 +1583,7 @@ func Cat(f Fs, w io.Writer, offset, count int64) error {
 }
 
 // Rcat reads data from the Reader until EOF and uploads it to a file on remote
-func Rcat(fdst Fs, dstFileName string, in0 io.ReadCloser, modTime time.Time) (err error) {
```
Comment (ncw) [inline fs/operations.go]: Why did you change the API here? returning the object created is useful isn't it?

Comment (breunigs) [reply fs/operations.go]: Hm, I'm not quite sure what you're referring to here. I'm actually extending the API to *also* return the `fs.Object` if possible. This was needed so that I can still attach it to the write handle in FUSE – previously `Rcat` would just return `error`.

Comment (ncw) [reply fs/operations.go]: Yes you are right - diff reading failure ;-) Sorry!
